### ISSUE TITLE
wrong_self_invocation violated for non-`self` `Self` types

### ISSUE THREAD

**Issue by Antikyth** [rust-lang/rust-clippy#10210]
### Summary

The `wrong_self_convention` lint is violated when a `to_*` function takes a parameter of type `&Self` when that parameter isn't an `&self` receiver or when an `into_*` function takes a parameter of type `Self`.

The [lint page] references the API guidelines for this lint, however those API guidelines only say that the parameter must be borrowed or owned etc., not that it must be in the  form `&self` or `self`.

[lint page]: https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention 

### Lint Name

wrong_self_invocation

### Reproducer

I tried this code (I cannot vouch for its quality, but nonetheless):

```rust
pub fn to_hex(Self(red, green, blue): &Self) -> u32 {
    let (red, green, blue) = (
        u32::from(red >> 8),
        u32::from(green >> 8),
        u32::from(blue >> 8),
    );
    
    (red << 16) | (green << 8) | blue
}
```

and the lint was violated:
```
methods called `to_*` usually take self by reference or self by value; consider choosing a less ambiguous name
```

but I wouldn't have expected that the lint would be violated, considering it takes a borrowed `Self` type.  One can certainly argue that the parameter should be written as `&self` instead, but that is not the point of this lint.

### Version

```text
rustc 1.68.0-nightly (bdb07a8ec 2022-12-11)
binary: rustc
commit-hash: bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4
commit-date: 2022-12-11
host: x86_64-unknown-linux-gnu
release: 1.68.0-nightly
LLVM version: 15.0.6
```

**Comment by Alexendoo** [comment]
Couldn't reproduce this myself, is this like what you have locally? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=095fc2038eae80a32d9b115fc51ef3e7

**Comment by Antikyth** [comment]
> Couldn't reproduce this myself, is this like what you have locally? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=095fc2038eae80a32d9b115fc51ef3e7

That, but with `u16` color values, yeah.  Odd.  Might be the specific version I'm using or something?

**Comment by Antikyth** [comment]
https://github.com/XdotRS/xrb/blob/main/src/common/visual.rs#L164-L183

This is the actual code as it is right now, just that I changed `Self(red, green, blue): &Self` to `&self` anyway, because the code read better that way... not sure if there is some other context that this helps find though.
